Fixing NME in #180 by fallback logics instead of additional information from chem template

rwxayheee opened the pull request:
This PR offers two types of fallback logics to handle systems with modified residues that aren't fully compliant with the template reaction and/or the expected adjacent molecule SMARTS. Additional information from chem template is not required. 

Here the fallback options are explained with NME as an example: 

(1) While making a padded NME

NME is the target residue molecule. It's only being padded on the N-term side. However, the template reaction smarts requires more than NME can offer as the reactant. 

In this case, the MCS (maximum common substructure) of NME and the template reactant will be computed. The non-common part of the reactant will be dropped. The MCS will be accepted as a fallback reactant as long as the required atom from adjacent molecule is present. The following information will be printed to standard output: 

```
target_mol ([H]NC([H])([H])[H]) is not fully compliant with the template rxn ([C:1](=[O:2])[C:3][N:4]>>[C:1](=[O:2])[C:3][N:4][C:11](=[O:12])[C:13])...
Switched from Template rxn ([C:1](=[O:2])[C:3][N:4]>>[C:1](=[O:2])[C:3][N:4][C:11](=[O:12])[C:13]) to Fallback rxn ([#6:3]-[#7:4]>>[C:3][N:4][C:11](=[O:12])[C:13])
```

(2) While making a padded residue with NME

NME is the adjacent residue molecule. It's used as source of atom positions to pad the C-term of its previous residue. However, the template adjacent res SMARTS requires more than NME can offer as the adjacent mol. 

In this case, only one fallback option will be considered which is to drop the unmapped atoms from adjacent res SMARTS, as these atoms won't be used as source of atom positions anyways. The following information will be printed to standard output: 

```
adjacent_mol ([H]NC([H])([H])[H]) is not fully compliant with the template adjacent_smarts (C(=O)[C:11][N:12])...
Switched from Template adjacent mol (C(=O)[C:11][N:12]) to Fallback adjacent mol ([C:11][N:12])
```

Apart from the fallback logics, the refactoring includes the following aspects: 

- Simplify react_and_map in rdkitutils
- Modularize existing functions in class ResiduePadder into smaller functions
- Fill in incomplete docstrings and type hints in the relavant parts
- Use list/set/dictionary comprehensions 
- Other refactoring/minor fixes to extend error handling
- Fix #182
- Minor fix to add to #183

Overall, I think this PR is better than what I previously suggested (https://github.com/forlilab/Meeko/pull/184#issue-2560408426) and can be useful for other nonstandard residues not just NME. The changes shouldn't change the behaviors of handling standard residues and it doesn't require further updates of chem templates (not sure why GitHub marked the diff..). 

@diogomart Could you please take a look at the changes at your convenience, and let me know what you think? 


As an example (original PDB file of 2G5Z): 
[2g5z.pdb.txt](https://github.com/user-attachments/files/17254630/2g5z.pdb.txt)
It has a nonstandard residue DHA at B:66 and a broken ILE at B:229. 

Command
```
mk_prepare_receptor.py --pdb 2g5z.pdb --delete_residues B:66,229 --output_filename 2g5z --skip_gpf
```

Standard Output
```
[00:13:35] product atom-mapping number 11 not found in reactants.
[00:13:35] product atom-mapping number 12 not found in reactants.
[00:13:35] product atom-mapping number 11 not found in reactants.
[00:13:35] product atom-mapping number 12 not found in reactants.
[00:13:35] product atom-mapping number 13 not found in reactants.
adjacent_mol ([H]NC([H])([H])[H]) is not fully compliant with the template adjacent_smarts (C(=O)[C:11][N:12])...
Switched from Template adjacent mol (C(=O)[C:11][N:12]) to Fallback adjacent mol ([C:11][N:12])
target_mol ([H]NC([H])([H])[H]) is not fully compliant with the template rxn ([C:1](=[O:2])[C:3][N:4]>>[C:1](=[O:2])[C:3][N:4][C:11](=[O:12])[C:13])...
Switched from Template rxn ([C:1](=[O:2])[C:3][N:4]>>[C:1](=[O:2])[C:3][N:4][C:11](=[O:12])[C:13]) to Fallback rxn ([#6:3]-[#7:4]>>[C:3][N:4][C:11](=[O:12])[C:13])
[00:13:35] product atom-mapping number 11 not found in reactants.
[00:13:35] product atom-mapping number 12 not found in reactants.
[00:13:35] product atom-mapping number 13 not found in reactants.

Files written:
2g5z.pdbqt <-- static (i.e., rigid) receptor input file
```

Output PDBQT: 
[2g5z.pdbqt.txt](https://github.com/user-attachments/files/17254729/2g5z.pdbqt.txt)

rwxayheee committed:
1. refactor ResiduePadder; 2. modify react_and_map in rdkitutils;
3. update chem templates for NME

rwxayheee committed:
provide fallback options for NME

rwxayheee committed:
complete fallback logics

rwxayheee committed:
move fallback logics to else

rwxayheee committed:
check padded_mol WIP

rwxayheee committed:
tar mol fallback logic WIP

rwxayheee committed:
sync chem templates

rwxayheee committed:
finished fallback logics

rwxayheee committed:
sync chem templates

rwxayheee committed:
beautify react_and_map

rwxayheee committed:
beautify class ResiduePadder and utility functions

rwxayheee committed:
fix typos

diogomart commented on meeko/linked_rdkit_chorizo.py:
```diff
@@ -496,15 +432,13 @@ def _delete_residues(res_to_delete, raw_input_mols):
         if res not in raw_input_mols:
             missing.add(res)
-        raw_input_mols.pop(res)
+        raw_input_mols.pop(res, None)
```
Nice!

rwxayheee commented:
This also works really well with modified nucleotides like PSU (Pseudouridine), which is present in ~800 PDB structures indexed by NAKB. PSU has a C-glycosidic bond instead of the typical N-glycosidic bond. Each of the two embedded PSU residues in PDB ID 8R8P will give: 
```
adjacent_mol ([H]OC1([H])C([H])(c2c([H])n([H])c(=O)n([H])c2=O)OC([H])(C([H])([H])O[PH](=O)[O-])C1([H])O) is not fully compliant with the template adjacent_smarts ([O&+0&X2&h1:11][C&X4:12]1CC(n)[O&X2]C1)...
Switched from Template adjacent mol ([O&+0&X2&h1:11][C&X4:12]1CC(n)[O&X2]C1) to Fallback adjacent mol ([O&+0&X2&h1:11][C&X4:12])
```
It might be helpful to also print the residue name and ID, but at the moment, the residue information isn't accessible at the point where the padder is called and the relevant molecules are evaluated.

diogomart commented:
This looks great. Thank you so much.

Currently having an issue with this pdb input
```
ATOM      1  CAY ACE P   0      -6.438  -3.380   5.569  1.00  0.00           C  
ATOM      2  CY  ACE P   0      -6.521  -2.744   4.224  1.00  0.00           C  
ATOM      3  OY  ACE P   0      -6.360  -3.405   3.200  1.00  0.00           O  
ATOM      4  HY1 ACE P   0      -6.612  -2.613   6.354  1.00  0.00           H  
ATOM      5  HY2 ACE P   0      -5.430  -3.827   5.706  1.00  0.00           H  
ATOM      6  HY3 ACE P   0      -7.209  -4.175   5.652  1.00  0.00           H  
ATOM      7  N   GLY P   1      -6.778  -1.424   4.200  1.00  0.00           N  
ATOM      8  CA  GLY P   1      -6.878  -0.708   2.896  1.00  0.00           C  
ATOM      9  C   GLY P   1      -5.557  -0.840   2.138  1.00  0.00           C  
ATOM     10  O   GLY P   1      -4.640  -1.504   2.579  1.00  0.00           O  
ATOM     11  HN  GLY P   1      -6.909  -0.881   5.026  1.00  0.00           H  
ATOM     12  HA1 GLY P   1      -7.046   0.342   3.095  1.00  0.00           H  
ATOM     13  HA2 GLY P   1      -7.677  -1.140   2.309  1.00  0.00           H  
```

which results in 
```
RuntimeError: target_mol does not contain the expected reactant with the expected target_required_atom_index
```

rwxayheee commented:
Hi @diogomart 
Thanks so much for the testing. I did one thing wrong with the reactant fallback. I will come up with a fix a little later today. A reactant fallback should pass for ACE, after dropping the unchanged N which ACE doesn't have

Specifically this was wrong: 
https://github.com/rwxayheee/Meeko/blob/9a5b6347e62ab9f945a217179c13d5909f4010ec/meeko/linked_rdkit_chorizo.py#L1892

rwxayheee committed:
fix handling of target_required_atom_index
during evaluation of fallback reactants

rwxayheee committed:
update error messages

rwxayheee committed:
update inline comments

rwxayheee commented:
Here's a minor change to the line mentioned in the previous comment: 

**Before**
```py
            # Filter fallback options by target_required_atom_index
            fallback_reactants = [
                reactant_mol for reactant_mol in apply_atom_mappings(fallback_reactant_smartsmol, reactant_smartsmol)
                if target_required_atom_index in get_molAtomMapNumbers(reactant_mol)
            ]
            if len(fallback_reactants) == 0:
                raise RuntimeError(f"target_mol does not contain the expected reactant with the expected target_required_atom_index.")
```

This was wrong. Not sure why it was written the way before. The comprehension includes too many things, I probably forgot what I was doing... 

**After**
```py
            # Add mapping number to fallback reactants and filter the fallback options
            # To be accepted, the fallback reactant needs to at least have a match with target_mol
            # containing target_mol's atom with target_required_atom_index
            fallback_reactants = [
                reactant_mol for reactant_mol in apply_atom_mappings(fallback_reactant_smartsmol, reactant_smartsmol)
                if any(target_required_atom_index in match for match in target_mol.GetSubstructMatches(reactant_mol))
            ]
            if len(fallback_reactants) == 0:
                raise RuntimeError(f"The maximum common substructure between target_mol and the expected reactant does not contain the expected linker atom with target_required_atom_index.")
```

To be accepted as a Fallback reactant, the candidate (which is the MCS of target_mol and Template reactant) just needs to have a possible match with target_mol that contains the linker atom, which can be checked by target_required_atom_index. 


The ACE example should be working now. Here's input & output: 

```
(mk_dev) amyhe@vpn-241-210 Downloads % mk_prepare_receptor.py --pdb ACE.pdb --output_filename ACE --skip_gpf                             
/Users/amyhe/micromamba/envs/mk_dev/bin/mk_prepare_receptor.py:4: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  __import__('pkg_resources').require('meeko==0.6.0a3')
templates=<meeko.linked_rdkit_chorizo.ResidueChemTemplates object at 0x161d80e80>
target_mol ([H]C([H])([H])C=O) is not fully compliant with the template rxn ([C:1](=[O:2])[C:3][N:4]>>[C:11][N:12][C:1](=[O:2])[C:3][N:4])...
Switched from Template rxn ([C:1](=[O:2])[C:3][N:4]>>[C:11][N:12][C:1](=[O:2])[C:3][N:4]) to Fallback rxn ([#6:1](=[#8:2])-[#6:3]>>[C:11][N:12][C:1](=[O:2])[C:3])
[20:03:11] product atom-mapping number 11 not found in reactants.
[20:03:11] product atom-mapping number 12 not found in reactants.
[20:03:11] product atom-mapping number 11 not found in reactants.
[20:03:11] product atom-mapping number 12 not found in reactants.
adjacent_mol ([H]C([H])([H])C=O) is not fully compliant with the template adjacent_smarts ([C:11](=[O:12])[C:13]N)...
Switched from Template adjacent mol ([C:11](=[O:12])[C:13]N) to Fallback adjacent mol ([C:11](=[O:12])[C:13])
[20:03:11] product atom-mapping number 11 not found in reactants.
[20:03:11] product atom-mapping number 12 not found in reactants.
[20:03:11] product atom-mapping number 13 not found in reactants.

Files written:
ACE.pdbqt <-- static (i.e., rigid) receptor input file
```
[ACE.pdb.txt](https://github.com/user-attachments/files/17286848/ACE.pdb.txt)
[ACE.pdbqt.txt](https://github.com/user-attachments/files/17286849/ACE.pdbqt.txt)

rwxayheee committed:
minor corrections to type hints

rwxayheee commented:
Made a minor correction to type hints in fb22563 
Small edit to error messages in 12a728d

rwxayheee committed:
small edit to error messages

diogomart reviewed (approved):
Works great, very nice design! Thanks :D

rwxayheee commented:
Thanks again for looking at the PR @diogomart. Merging to develop
